<H1>First part of the Task</H1>

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd, numpy as np
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data  = requests.get(url).text 
soup = BeautifulSoup(data,"html5lib")
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})


In [2]:
df.shape

(103, 3)

There is 103 rows and 3 columns

<H1>Second part of the Task</H1>

In [3]:
lat_long = pd.read_csv('Geospatial_Coordinates.csv')  
lat_long

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [4]:
toronto = df.sort_values(['PostalCode'])
toronto.reset_index(inplace=True)
toronto.drop(['index'], axis=1,inplace=True)

In [5]:
toronto['Latitude'] = lat_long['Latitude']
toronto['Longitude'] = lat_long['Longitude']

<H3> Result Dataframe </H3>

In [6]:
toronto

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

<H1>Third part of the Task</H1>

In [7]:
import config
import folium
from geopy.geocoders import Nominatim
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

In [8]:
CLIENT_ID = 'WJQXGWEEU2F5M2CFKLZ1H4TTSYNZ5OWU5MCFFXAAJCWXKZC2' # your Foursquare ID
CLIENT_SECRET = 'ZYDU55ECKA1ULZXXCYIIWXIDHRLTBFPA04PG1TJOWHA24VA1' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

In [9]:
#Get the coordinates of Toronto
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


<H3>Map of Toronto and the Neighborhoods</H3>

In [10]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<H3>Venues of each Neighborhood</H3>

In [11]:
#Function to get venues for each neighborhood
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


toronto_venues = getNearbyVenues(toronto['Neighborhood'], toronto['Latitude'], toronto['Longitude'])
toronto_venues

Neighborhood  \
0                                        Malvern, Rouge   
1                Rouge Hill, Port Union, Highland Creek   
2                Rouge Hill, Port Union, Highland Creek   
3                     Guildwood, Morningside, West Hill   
4                     Guildwood, Morningside, West Hill   
...                                                 ...   
2105  South Steeles, Silverstone, Humbergate, Jamest...   
2106  Clairville, Humberwood, Woodbine Downs, West H...   
2107  Clairville, Humberwood, Woodbine Downs, West H...   
2108  Clairville, Humberwood, Woodbine Downs, West H...   
2109  Clairville, Humberwood, Woodbine Downs, West H...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.806686              -79.194353                 Wendy’s   
1                 43.784535              -79.160497  Chris Effects Painting   
2                 43.784535              -79.160497   Royal Canadian Legion   
3                 43.763573              -79.188711          RBC Royal Bank   
4                 43.763573              -79.188711       G & G Electronics   
...                     ...                     ...                     ...   
2105              43.739416              -79.588437               Dollarama   
2106              43.706748              -79.594054      Economy Rent A Car   
2107              43.706748              -79.594054  Logistics Distribution   
2108              43.706748              -79.594054           Saand Rexdale   
2109              43.706748              -79.594054    Vectra Heavy Haulers   

      Venue Latitude  Venue Longitude              Venue Category  
0          43.807448       -79.199056        Fast Food Restaurant  
1          43.784343       -79.163742  Construction & Landscaping  
2          43.782533       -79.163085                         Bar  
3          43.766790       -79.191151                        Bank  
4          43.765309       -79.191537           Electronics Store  
...              ...              ...                         ...  
2105       43.742018       -79.591076              Discount Store  
2106       43.708471       -79.589943         Rental Car Location  
2107       43.707554       -79.589252                         Bar  
2108       43.705072       -79.598725                   Drugstore  
2109       43.704891       -79.599410                  Truck Stop  

[2110 rows x 7 columns]

In [12]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
neig = toronto_onehot['Neighborhood']
toronto_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
toronto_onehot.insert(0, 'Neighborhood', neig)
toronto_onehot

Neighborhood  Accessories Store  \
0                                        Malvern, Rouge                  0   
1                Rouge Hill, Port Union, Highland Creek                  0   
2                Rouge Hill, Port Union, Highland Creek                  0   
3                     Guildwood, Morningside, West Hill                  0   
4                     Guildwood, Morningside, West Hill                  0   
...                                                 ...                ...   
2105  South Steeles, Silverstone, Humbergate, Jamest...                  0   
2106  Clairville, Humberwood, Woodbine Downs, West H...                  0   
2107  Clairville, Humberwood, Woodbine Downs, West H...                  0   
2108  Clairville, Humberwood, Woodbine Downs, West H...                  0   
2109  Clairville, Humberwood, Woodbine Downs, West H...                  0   

      Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0           0                   0             0               0   
1           0                   0             0               0   
2           0                   0             0               0   
3           0                   0             0               0   
4           0                   0             0               0   
...       ...                 ...           ...             ...   
2105        0                   0             0               0   
2106        0                   0             0               0   
2107        0                   0             0               0   
2108        0                   0             0               0   
2109        0                   0             0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
2105                0                 0                    0             0   
2106                0                 0                    0             0   
2107                0                 0                    0             0   
2108                0                 0                    0             0   
2109                0                 0                    0             0   

      ...  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0     ...                              0                 0            0   
1     ...                              0                 0            0   
2     ...                              0                 0            0   
3     ...                              0                 0            0   
4     ...                              0                 0            0   
...   ...                            ...               ...          ...   
2105  ...                              0                 0            0   
2106  ...                              0                 0            0   
2107  ...                              0                 0            0   
2108  ...                              0                 0            0   
2109  ...                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0                         0                0         0          0   
1                         0                0         0          0   
2                         0                0         0          0   
3                         0                0         0          0   
4                         0                0         0          0   
...                     ...              ...     

In [13]:
#Information grouped by Neighborhood
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
94                                 Willowdale South                0.0   
95                                  Willowdale West                0.0   
96                                           Woburn                0.0   
97                                 Woodbine Heights                0.0   
98                                  York Mills West                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
1       0.0                 0.0           0.0             0.0   
2       0.0                 0.0           0.0             0.0   
3       0.0                 0.0           0.0             0.0   
4       0.0                 0.0           0.0             0.0   
..      ...                 ...           ...             ...   
94      0.0                 0.0           0.0             0.0   
95      0.0                 0.0           0.0             0.0   
96      0.0                 0.0           0.0             0.0   
97      0.0                 0.0           0.0             0.0   
98      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0                 0.00           0.0  ...   
1               0.0               0.0                 0.00           0.0  ...   
2               0.0               0.0                 0.00           0.0  ...   
3               0.0               0.0                 0.00           0.0  ...   
4               0.0               0.0                 0.04           0.0  ...   
..              ...               ...                  ...           ...  ...   
94              0.0               0.0                 0.00           0.0  ...   
95              0.0               0.0                 0.00           0.0  ...   
96              0.0               0.0                 0.00           0.0  ...   
97              0.0               0.0                 0.00           0.0  ...   
98              0.0               0.0                 0.00           0.0  ...   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                             0.0               0.0     0.000000   
1                             0.0               0.0     0.000000   
2                             0.0               0.0     0.000000   
3                             0.0               0.0     0.000000   
4                             0.0               0.0     0.000000   
..                            ...               ...          ...   
94                            0.0               0.0     0.000000   
95                            0.0               0.0     0.000000   
96                            0.0               0.0     0.000000   
97                            0.0               0.0     0.142857   
98                            0.0               0.0     0.000000   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                0.000000              0.0       0.0        0.0          0.0   
1                0.000000              0.0       0.0        0.0          0.0   
2                0.000000              0.0       0.0        0.0          0.0   
3                0.000000              0.0       0.0        0.0          0.0   
4                0.000000              0.0       0.0        0.0          0.0   
..                    ...              ...       ...        ...          ...   
94               0.

In [14]:
#New dataframe that display the top 10 venues for each neighborhood.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Agincourt                Lounge   
1                            Alderwood, Long Branch           Pizza Place   
2   Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                   Bayview Village                  Café   
4                 Bedford Park, Lawrence Manor East    Italian Restaurant   
..                                              ...                   ...   
94                                 Willowdale South           Coffee Shop   
95                                  Willowdale West              Pharmacy   
96                                           Woburn           Coffee Shop   
97                                 Woodbine Heights              Bus Stop   
98                                  York Mills West                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Latin American Restaurant        Breakfast Spot        Clothing Store   
1                          Gym        Sandwich Place           Coffee Shop   
2                         Bank        Sandwich Place           Supermarket   
3          Japanese Restaurant                  Bank    Chinese Restaurant   
4               Sandwich Place           Coffee Shop            Hobby Shop   
..                         ...                   ...                   ...   
94            Ramen Restaurant         Shopping Mall      Sushi Restaurant   
95                 Coffee Shop           Supermarket               Butcher   
96       Korean BBQ Restaurant     Other Repair Shop   Dumpling Restaurant   
97                        Park            Beer Store          Skating Rink   
98  Construction & Landscaping     Convenience Store           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Skating Rink   Dumpling Restaurant          Distribution Center   
1             Playground                   Pub                      Dog Run   
2          Deli / Bodega      Sushi Restaurant            Mobile Phone Shop   
3     Dim Sum Restaurant        Discount Store          Distribution Center   
4                Butcher                   Pub            Indian Restaurant   
..                   ...                   ...                          ...   
94                  Café           Pizza Place                   Restaurant   
95           Pizza Place     Electronics Store  Eastern European Restaurant   
96        Discount Store   Distribution Center                      Dog Run   
97           Video Store    Athletics & Sports                  Curling Ice   
98             Drugstore        Discount Store          Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Dog Run      Doner Restaurant             Donut Shop  
1     Dim Sum Restaurant                 Diner         Discount Store  
2         Ice Cream Shop                 Diner    Fried Chicken Joint  
3                Dog Run      Doner Restaurant             Donut Shop  
4       Sushi Restaurant            Restaurant              Juice Bar  
..                   ...                   ...                    ...  
94  Fast Food Restaurant                  Bank         Ice Cream Shop  
95   Dumpling Restaurant             Drugstore       Department Store  
96      Doner Restaurant            Donut Shop              Drugstore  
97            Donut Shop        Discount Store    Distribution Center  
98               Dog Run      Doner Restaurant             Donut Shop  

[99 rows x 11 columns]

## Cluster Neighborhoods


In [15]:
# set number of clusters
kclusters = 10

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(50)

PostalCode                 Borough  \
0         M1B             Scarborough   
1         M1C             Scarborough   
2         M1E             Scarborough   
3         M1G             Scarborough   
4         M1H             Scarborough   
5         M1J             Scarborough   
6         M1K             Scarborough   
7         M1L             Scarborough   
8         M1M             Scarborough   
9         M1N             Scarborough   
10        M1P             Scarborough   
11        M1R             Scarborough   
12        M1S             Scarborough   
13        M1T             Scarborough   
14        M1V             Scarborough   
15        M1W             Scarborough   
16        M1X             Scarborough   
17        M2H              North York   
18        M2J              North York   
19        M2K              North York   
20        M2L              North York   
21        M2M              North York   
22        M2N              North York   
23        M2P              North York   
24        M2R              North York   
25        M3A              North York   
26        M3B              North York   
27        M3C              North York   
28        M3H              North York   
29        M3J              North York   
30        M3K              North York   
31        M3L              North York   
32        M3M              North York   
33        M3N              North York   
34        M4A              North York   
35        M4B               East York   
36        M4C               East York   
37        M4E            East Toronto   
38        M4G               East York   
39        M4H               East York   
40        M4J  East York/East Toronto   
41        M4K            East Toronto   
42        M4L            East Toronto   
43        M4M            East Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
46        M4R         Central Toronto   
47        M4S         Central Toronto   
48        M4T         Central Toronto   
49        M4V         Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                      Malvern, Rouge  43.806686 -79.194353   
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                         Birch Cliff, Cliffside West  43.692657 -79.264848   
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                  Wexford, Maryvale  43.750072 -79.295849   
12                                          Agincourt  43.794200 -79.262029   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   
16                                        Upper Rouge  43.836125 -79.205636   
17                                  Hillcrest Village  43.803762 -79.363452   
18                       Fairview, Henry Farm, Oriole  43.778517 -79.346556   
19                                    Bayview Village  43.786947 -79.385975   
20                           York Mills, Silver Hills  43.757490 -79.374714   
21                            Willowdale, Newtonbrook  43.789053 -79.408493   
22                                   Willowdale South  43.770120 -79.408493   
23             

In [16]:
#Drop Neighborhood without values
toronto_merged = toronto_merged.dropna()

#Change clusters type from float to int
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype('int')

In [17]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
map_clusters       

In [18]:
#Examine Cluster 0
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2            Scarborough               0            Donut Shop   
3            Scarborough               0           Coffee Shop   
4            Scarborough               0                  Bank   
6            Scarborough               0      Department Store   
7            Scarborough               0                Bakery   
..                   ...             ...                   ...   
95             Etobicoke               0              Pharmacy   
96            North York               0                   Gym   
99             Etobicoke               0           Pizza Place   
101            Etobicoke               0         Grocery Store   
102  Etobicoke Northwest               0            Truck Stop   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2           Medical Center            Restaurant   Rental Car Location   
3    Korean BBQ Restaurant     Other Repair Shop   Dumpling Restaurant   
4         Hakka Restaurant   Fried Chicken Joint       Thai Restaurant   
6           Discount Store           Coffee Shop            Hobby Shop   
7                 Bus Line         Metro Station          Intersection   
..                     ...                   ...                   ...   
95              Beer Store             Pet Store           Pizza Place   
96             Yoga Studio   Dumpling Restaurant        Discount Store   
99          Sandwich Place            Playground           Coffee Shop   
101            Pizza Place              Pharmacy            Beer Store   
102    Rental Car Location                   Bar             Drugstore   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                    Bank    Mexican Restaurant          Intersection   
3          Discount Store   Distribution Center               Dog Run   
4    Caribbean Restaurant    Athletics & Sports           Gas Station   
6      Chinese Restaurant             Drugstore   Distribution Center   
7          Ice Cream Shop           Bus Station          Soccer Field   
..                    ...                   ...                   ...   
95            Coffee Shop                  Café          Liquor Store   
96    Distribution Center               Dog Run      Doner Restaurant   
99         Discount Store          Intersection    Chinese Restaurant   
101        Sandwich Place        Discount Store   Fried Chicken Joint   
102           Yoga Studio        Discount Store   Distribution Center   

    8th Most Common Venue        9th Most Common Venue  \
2          Breakfast Spot            Electronics Store   
3        Doner Restaurant                   Donut Shop   
4                  Bakery                      Dog Run   
6                 Dog Run             Doner Restaurant   
7                    Park          Dumpling Restaurant   
..                    ...                          ...   
95              Gift Shop            German Restaurant   
96             Donut Shop                    Drugstore   
99            Yoga Studio                      Dog Run   
101  Fast Food Restaurant  Eastern European Restaurant   
102               Dog Run             Doner Restaurant   

          10th Most Common Venue  
2            Dumpling Restaurant  
3                      Drugstore  
4                 Discount Store  
6                     Donut Shop  
7                      Drugstore  
..                           ...  
95   Eastern European Restaurant  
96   Eastern European Restaurant  
99            Dim Sum Restaurant  
101          Dumpling Restaurant  
102                   Donut Shop  

[76 rows x 12 columns]

In [19]:
#Examine Cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1               Scarborough               1  Construction & Landscaping   
14              Scarborough               1                Intersection   
23               North York               1                        Park   
25               North York               1           Food & Drink Shop   
40   East York/East Toronto               1                        Park   
44          Central Toronto               1                        Park   
50         Downtown Toronto               1                        Park   
64          Central Toronto               1                        Park   
73                     York               1                        Park   
74                     York               1                        Park   
79               North York               1                        Park   
98                     York               1                        Park   
100               Etobicoke               1                        Park   

          2nd Most Common Venue       3rd Most Common Venue  \
1                           Bar                 Yoga Studio   
14                         Park                  Playground   
23   Construction & Landscaping           Convenience Store   
25                         Park  Construction & Landscaping   
40                Metro Station           Convenience Store   
44                     Bus Line                 Swim School   
50                   Playground                       Trail   
64                        Trail            Sushi Restaurant   
73                        Field                Hockey Arena   
74                Women's Store                        Pool   
79                       Bakery  Construction & Landscaping   
98            Convenience Store                 Yoga Studio   
100              Sandwich Place                    Bus Line   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1     Dumpling Restaurant   Distribution Center               Dog Run   
14             Donut Shop    Dim Sum Restaurant                 Diner   
23            Yoga Studio             Drugstore        Discount Store   
25            Yoga Studio        Discount Store   Distribution Center   
40            Yoga Studio        Discount Store   Distribution Center   
44       Business Service            Donut Shop        Discount Store   
50            Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
64          Jewelry Store           Escape Room     Electronics Store   
73                  Trail           Yoga Studio            Donut Shop   
74              Drugstore                 Diner        Discount Store   
79       Basketball Court           Yoga Studio   Dumpling Restaurant   
98              Drugstore        Discount Store   Distribution Center   
100     Mobile Phone Shop           Yoga Studio      Doner Restaurant   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1               Doner Restaurant            Donut Shop             Drugstore   
14                Discount Store   Distribution Center               Dog Run   
23           Distribution Center               Dog Run      Doner Restaurant   
25                       Dog Run      Doner Restaurant            Donut Shop   
40                       Dog Run      Doner Restaurant            Donut Shop   
44           Distribution Center               Dog Run      Doner Restaurant   
50                         Diner        Discount Store   Distribution Center   
64   Eastern European Restaurant   Dumpling Restaurant             Drugstore   
73                         Diner        Discount Store   Distribution Center   
74           Distribution Center               Dog Run      Doner Restaurant   
79                       Dog Run      Doner Restaurant            Donut Shop   
98                       Dog Run      Doner Restaurant            Donut Shop   
100               Disco

In [20]:
#Examine Cluster 4
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
30       North York               4               Airport   
48  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
30                  Park           Yoga Studio             Drugstore   
48           Yoga Studio             Drugstore                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
30                 Diner        Discount Store   Distribution Center   
48        Discount Store   Distribution Center               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
30               Dog Run      Doner Restaurant             Donut Shop  
48      Doner Restaurant            Donut Shop    Dumpling Restaurant